torch/csrc/jit/script/compiler.cpp

```diff
@@ -34,23 +34,127 @@ using ListAttributeMap = std::unordered_map<
     std::string,
     std::pair<const std::vector<double>, std::string>>;
 
+// Keep track of environment as we descend down nested control
+// structures.
+struct Environment {
+  Environment(Block* b, std::shared_ptr<Environment> next = nullptr)
+      : b(b), next(next) {}
+
+  std::vector<std::string> captured_inputs;
+  ValueTable value_table;
+  Block* b;
+
+  std::shared_ptr<Environment> next;
+
+  Value* findInThisFrame(const std::string& name) {
+    if (value_table.count(name)) {
+      return value_table.at(name);
+    }
+    return nullptr;
+  }
+
+  Value* findInParentFrame(const std::string& name) {
+    for (auto runner = next; runner; runner = runner->next) {
+      if (runner->value_table.count(name)) {
+        return runner->value_table.at(name);
+      }
+    }
+    return nullptr;
+  }
+
+  Value* createCapturedInput(const std::string& name) {
+    // Create the input
+    Value* new_input = b->addInput();
+
+    // Associate this name with this value
+    value_table[name] = new_input;
+
+    // List as a positional input
+    captured_inputs.push_back(name);
+
+    return new_input;
+  }
+
+  Symbol getBlockOwningKind() {
+    Symbol owning_kind = Symbol();
+    if (b->owningNode()) {
+      owning_kind = b->owningNode()->kind();
+    }
+    return owning_kind;
+  }
+
+  void setVar(const std::string& name, Value* value) {
+    if (!findInThisFrame(name) && findInParentFrame(name) &&
+        getBlockOwningKind() == Symbol("Loop"))
+      createCapturedInput(name);
+    value_table[name] = value;
+  }
+
+  Value* getVar(const Ident& ident) {
+    return getVar(ident.name(), ident);
+  }
+
+  Value* getVar(const std::string& ident, const TreeView& tv) {
+    Value* retval = findInThisFrame(ident);
+
+    if (!retval && (retval = findInParentFrame(ident)) &&
+        getBlockOwningKind() == Symbol("Loop"))
+      retval = createCapturedInput(ident);
+
+    if (!retval) {
+      throw ErrorReport(tv) << "undefined value " << ident;
+    }
+
+    return retval;
+  }
+
+  // Given that after emitting statements in a block, we've added block inputs
+  // for all value references and assignments, delete inputs for which there was
+  // no assignment, only references.
+  void deleteExtraInputs(size_t skip_num = 0) {
```

Comment:
I don't understand why do we need this. The whole contract of `getVar` + `deleteExtraInputs` is a bit unclear to me. Since we never want to add values we only retrieve in the body of the loop, then why do we have `Loop` special cased in `getVar` and not in `setVar` (i.e. lazily add a mutated input if it's not in the current scope). That would avoid the need for `deleteExtraInputs`

Reply:
@zdevito  can articulate this better than I can

Reply:
In any case, having the example where this fires in the code is very helpful.

Reply:
Alright, so basically it boils down to replaceAllUsesWith not being constrained to nodes within a block. If I lazily add an input on mutation, I have to traverse all predecessor nodes within that block and manually replace references to that Value, whereas by pessimistically adding inputs for all references, we can just call `replaceAllUsesWith` with the Value* pointed to in the parent Environment

let me know if that makes sense

Reply:
That makes a lot of sense, but would deserve a comment!